### PR TITLE
port ssh util from vsc-manage to be reusable (HPC-7591)

### PR DESCRIPTION
**Pull request by wdpypere**
- cleaned it up
   - removed code that are now paramiko features (e.g. timeouts)
   - reversed code (manage has commands which might be ssh commands which is not very intuitive)
 - added 2 new features:
   - add ssh agent forwarding
   - read proxy from ssh config file

### CODE REVIEW

File: lib/vsc/utils/ssh_util.py [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+        self.client.set_missing_host_key_policy(paramiko.AutoAddPolicy())
+        self.client.load_system_host_keys()
+
+        try:
+            if self.passwd:
+                self.client.connect(hostname=self.host, username=self.user, timeout=self.conn_timeout,
+                                    sock=self.proxy, allow_agent=False)
+            else:
+                self.client.connect(hostname=self.host, timeout=self.conn_timeout, sock=self.proxy)
+
```
Comment (itkovian) [inline lib/vsc/utils/ssh_util.py]: `return True` here? Not sure.

File: lib/vsc/utils/ssh_util.py [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+
+        except paramiko.AuthenticationException:
+            logging.error("Authentication to host %s failed, please verify your credentials.", self.host)
+            self.disconnect()
```
Comment (itkovian) [inline lib/vsc/utils/ssh_util.py]: Is there actually a connection at this point?

File: lib/vsc/utils/ssh_util.py [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+                except Exception as ex:
+                    logging.error('Could not run command %s. error: %s', command, ex)
+                    exit_status = 1
+
+                try:
+                    ssh_stdout = stdout.read().strip()
+                    ssh_stderr = stderr.read().strip()
+
+                except Exception as ex:
+                    logging.error('Could not read from stdout/stderr. error: %s', ex)
```
Comment (itkovian) [inline lib/vsc/utils/ssh_util.py]: The logic is false, actually. You cannot read from at least one of them (the first one that fails), but setting stdout to "" may be misleading, as it actually could contain something.

Comment (itkovian) [reply lib/vsc/utils/ssh_util.py]: Make it two distinct try/except clauses.

File: lib/vsc/utils/ssh_util.py [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+                    exit_status = 1
+                    ssh_stdout = ""
+                    ssh_stderr = ex
+
+                if exit_status != 0:
+                    logging.error("Command failed: %s. stdout: %s stderr: %s",
+                                  command, ssh_stdout, ssh_stderr)
+                    self.disconnect()
+                    return exit_status, ssh_stderr
```
Comment (itkovian) [inline lib/vsc/utils/ssh_util.py]: You are returning different things compared to the below. Here you return an integer and an Exception. Below you return an integer and a list (output). This seems odd and error prone.

File: lib/vsc/utils/ssh_util.py [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+
+                else:
+                    output.append(ssh_stdout)
+
+            self.disconnect()
+            return 0, output
+
+        else:
+            logging.error("Could not establish SSH connection")
+            return 1, 'Could not establish SSH connection'
```
Comment (itkovian) [inline lib/vsc/utils/ssh_util.py]: And here you return an integer and a string :)

File: lib/vsc/utils/ssh_util.py (hunk 2 above)
Comment (wdpypere) [reply lib/vsc/utils/ssh_util.py]: not sure, probably not. Can't hurt to be safe?

File: lib/vsc/utils/ssh_util.py (hunk 3 above)
Comment (wdpypere) [reply lib/vsc/utils/ssh_util.py]: done

**Comment by wdpypere** [comment]
@itkovian remarks addressed

File: lib/vsc/utils/ssh_util.py (hunk 2 above)
Comment (itkovian) [reply lib/vsc/utils/ssh_util.py]: If it does not fail hard when there is no connection, then it should be ok :)

File: lib/vsc/utils/ssh_util.py (hunk 1 above)
Comment (itkovian) [reply lib/vsc/utils/ssh_util.py]: I know, but returning early is easier to see in this case. You do not do anything after the except clauses, so might as well return here.

Comment (itkovian) [reply lib/vsc/utils/ssh_util.py]: And then the return at the end can be False, avoiding all the separate returns in each except.

File: lib/vsc/utils/ssh_util.py [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+                except Exception as ex:
+                    logging.error('Could not run command %s. error: %s', command, ex)
+                    exit_status = 1
+
+                try:
+                    ssh_stdout = stdout.read().strip()
+                except Exception as ex:
+                    logging.error('Could not read from stdout. error: %s', ex)
+                    exit_status = 1
+                    ssh_stdout = ex
```
Comment (itkovian) [inline lib/vsc/utils/ssh_util.py]: Should be `ex.str()`? I am still not sure what the idea is with returning the exception.

File: lib/vsc/utils/ssh_util.py [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+
+                try:
+                    ssh_stderr = stderr.read().strip()
+                except Exception as ex:
+                    logging.error('Could not read from stderr. error: %s', ex)
+                    exit_status = 1
+                    ssh_stderr = ex
```
Comment (itkovian) [inline lib/vsc/utils/ssh_util.py]: Likewise, `ex.str()`?

File: lib/vsc/utils/ssh_util.py [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+"""
+Ssh helper class
+"""
+
+import os
+import logging
+import paramiko
+import socket
+
+class Ssh_util(object):
```
Comment (stdweird) [inline lib/vsc/utils/ssh_util.py]: use https://github.com/cwood/sshed ?

Comment (wdpypere) [reply lib/vsc/utils/ssh_util.py]: I'll look at fabric

**Pull request closed without merging**